### ISSUE TITLE
[Bug]: A5 2-D reduction over [TPT, K_POW2] drops lanes when K_POW2 < 8 (MoE histogram undercount)

### ISSUE THREAD

**Issue by zheliuyu** [triton-lang/triton-ascend#2247]
### Summary

On **Ascend A5** (Ascend950PR, `Ascend950PR_9579`), **reductions over a 2-D tile** loaded as `[TOKENS_PER_TILE, K_POW2]` produce **incorrect results** when the **inner constexpr extent `K_POW2` is 2 or 4** (i.e. `< 8`), even though:

- The **same indices** counted with a **1-D flatten + `tl.atomic_add`** path sum to the correct total `T×K`.
- **2-D `tl.load` / store dumps** match host memory (loads are correct; reductions are wrong).
- Padding the inner `tl.arange` width to **8** while keeping `k < K` masks restores correct counts.

This is a **runtime codegen / lowering bug**, not a compile failure. It affects both **`tl.atomic_add`** and **`tl.sum`** on tensors originating from the 2-D block load. **`tl.reshape` to 1-D does not fix** the bug—the reduction still behaves as if only a subset of lanes contributed.

Typical MoE routing case (Mixtral-style): `T=512`, `K=2`, `E=8`, production `TOKENS_PER_TILE=512`, `K_POW2=2` → histogram total **`256` instead of `1024`**, breaking `compute_routing_metadata` and downstream gather/scatter (~75% expert forward mismatch).

The same 2-D pattern with **`K_POW2 ≥ 8`** (e.g. pytest shape `T=256, E=32, K=8`) counts correctly on A5. **Ascend A3** passes the `K=2` production histogram in our downstream tests; the failure is **A5-specific**.

**Related (optional separate issue):** bf16 `tl.sum(..., axis=0)` on `[K, H]` gather tiles—see `triton-ascend-issue-moe-05-bf16-sum-axis0.md`.

### Minimal complete example

Save as `repro_moe_histogram.py` and run on an idle A5 NPU:

```bash
python repro_moe_histogram.py
```

```python
#!/usr/bin/env python3
"""Minimal A5: 2-D block reduction drops lanes when inner K_POW2 < 8."""

import sys

import torch
import triton
import triton.language as tl

T, E, K, K_POW2, TPT = 512, 8, 2, 2, 512


def host_ref(idx, e):
    return int(torch.bincount(idx.reshape(-1).clamp(0, e - 1), minlength=e).sum())


@triton.jit
def hist_1d_atomic(topk_ptr, partial_ptr, T, E: tl.constexpr, n_tiles, TPT: tl.constexpr,
                   K: tl.constexpr, K_POW2: tl.constexpr, E_POW2: tl.constexpr):
    tile = tl.program_id(0)
    e_offs = tl.arange(0, E_POW2)
    tl.store(partial_ptr + e_offs * n_tiles + tile, tl.zeros([E_POW2], tl.int32), mask=e_offs < E)
    n_flat: tl.constexpr = TPT * K_POW2
    flat = tl.arange(0, n_flat)
    tok = tile * TPT + flat // K_POW2
    kk = flat % K_POW2
    mask = (tok < T) & (kk < K)
    ids = tl.load(topk_ptr + tok * K + tl.minimum(kk, K - 1), mask=mask, other=-1)
    valid = mask & (ids >= 0) & (ids < E)
    tl.atomic_add(partial_ptr + tl.where(valid, ids, 0) * n_tiles + tile,
                  tl.full([n_flat], 1, dtype=tl.int32), mask=valid)


@triton.jit
def hist_2d_sum(topk_ptr, counts_ptr, T, E: tl.constexpr, n_tiles, TPT: tl.constexpr,
                K: tl.constexpr, K_POW2: tl.constexpr):
    tile, e = tl.program_id(0), tl.program_id(1)
    tok = tile * TPT + tl.arange(0, TPT)
    k = tl.arange(0, K_POW2)
    mask = (tok[:, None] < T) & (k[None, :] < K)
    ids = tl.load(topk_ptr + tok[:, None] * K + tl.minimum(k, K - 1)[None, :], mask=mask, other=-1)
    tl.store(counts_ptr + e * n_tiles + tile, tl.sum(tl.where((ids == e) & mask, 1, 0)))


@triton.jit
def hist_2d_atomic(topk_ptr, partial_ptr, T, E: tl.constexpr, n_tiles, TPT: tl.constexpr,
                   K: tl.constexpr, E_POW2: tl.constexpr, INNER_POW2: tl.constexpr):
    tile = tl.program_id(0)
    e_offs = tl.arange(0, E_POW2)
    tl.store(partial_ptr + e_offs * n_tiles + tile, tl.zeros([E_POW2], tl.int32), mask=e_offs < E)
    tok = tile * TPT + tl.arange(0, TPT)
    k = tl.arange(0, INNER_POW2)
    mask = (tok[:, None] < T) & (k[None, :] < K)
    ids = tl.load(topk_ptr + tok[:, None] * K + tl.minimum(k, K - 1)[None, :], mask=mask, other=-1)
    flat_e = tl.reshape(ids, [TPT * INNER_POW2])
    flat_m = tl.reshape(mask, [TPT * INNER_POW2])
    valid = flat_m & (flat_e >= 0) & (flat_e < E)
    tl.atomic_add(partial_ptr + tl.where(valid, flat_e, 0) * n_tiles + tile,
                  tl.full([TPT * INNER_POW2], 1, dtype=tl.int32), mask=valid)


def main():
    torch.npu.set_device(0)
    print("device:", torch.npu.get_device_name(0))
    n_tiles = triton.cdiv(T, TPT)
    e2 = triton.next_power_of_2(E)
    torch.manual_seed(0)
    idx = torch.topk(torch.randn(T, E, device="npu"), K, dim=-1).indices.to(torch.int32)
    ref = host_ref(idx, E)
    print(f"T={T} K={K} TPT={TPT} ref={ref}")

    p = torch.empty(E, n_tiles, dtype=torch.int32, device="npu")
    hist_1d_atomic[(n_tiles,)](idx, p, T, E=E, n_tiles=n_tiles, TPT=TPT, K=K, K_POW2=K_POW2, E_POW2=e2)
    torch.npu.synchronize()
    print("control_1d_atomic:", int(p.sum()), ref)

    c = torch.zeros(E, n_tiles, dtype=torch.int32, device="npu")
    hist_2d_sum[(n_tiles, E)](idx, c, T, E=E, n_tiles=n_tiles, TPT=TPT, K=K, K_POW2=K_POW2)
    torch.npu.synchronize()
    print("fail_2d_sum:", int(c.sum()), ref)

    p.zero_()
    hist_2d_atomic[(n_tiles,)](idx, p, T, E=E, n_tiles=n_tiles, TPT=TPT, K=K, E_POW2=e2, INNER_POW2=2)
    torch.npu.synchronize()
    print("fail_2d_atomic_inner2:", int(p.sum()), ref)

    p.zero_()
    hist_2d_atomic[(n_tiles,)](idx, p, T, E=E, n_tiles=n_tiles, TPT=TPT, K=K, E_POW2=e2, INNER_POW2=8)
    torch.npu.synchronize()
    print("control_pad_inner8:", int(p.sum()), ref)


if __name__ == "__main__":
    sys.exit(main())
```

### Expected vs observed

**Expected**

- Every variant counts **`T × K = 1024`** expert assignments (for the fixed seed / random top-k indices).

**Observed (A5, Ascend950PR_9579)**

```
device: Ascend950PR_9579
T=512 E=8 K=2 K_POW2=2 TPT=512 n_tiles=1 ref=1024
control_1d_atomic: got=1024 ref=1024 OK
fail_2d_sum: got=256 ref=1024 FAIL
fail_2d_atomic_kpow2_2: got=256 ref=1024 FAIL
control_pad_k_inner_8: got=1024 ref=1024 OK
```

Compiler may emit (informational):

```
[WARNING] AutoBlockify disabled for kernel ... Unsafe ops: atomic operations.
```

1-D atomic kernels show the same warning but produce **correct** totals; AutoBlockify is not the root cause.

### Bisection matrix (A5)

| Variant | `K=2`, `TPT=512`, `T=512` | Notes |
|--------|---------------------------|--------|
| 2-D load dump vs host | OK | Loads correct |
| 1-D flatten + `atomic_add` | **1024** | Control |
| 2-D load + `tl.sum` per expert | **256** | Not atomic-specific |
| 2-D load + reshape + 1-D `atomic_add` (`K_POW2=2`) | **256** | Reshape does not help |
| 2-D `atomic_add` without reshape (`K_POW2=2`) | **256** | Same undercount |
| 2-D path, **`INNER_POW2=8`**, mask `k < K` | **1024** | Workaround |
| `K=8`, `K_POW2=8`, various `TPT` | **OK** | Explains passing pytest `[256-32-8]` |

Extended bisection: `min_repro_moe_hist.py` with `KIND=prod|atomic|padk|kdim|load`.

### Downstream impact

In [Liger-Kernel](https://github.com/linkedin/Liger-Kernel) Ascend **fused MoE**, `_moe_router_histogram_kernel` uses 2-D loads with `K_POW2 = next_power_of_2(K)`. Mixtral (`E=8`, `K=2`) → `K_POW2=2`, `TOKENS_PER_TILE=512` → **`freq_sum=256`**, corrupt routing metadata, MixtralExperts ~75% mismatch. Fix in application code: pad inner histogram dim to **`max(8, K_POW2)`** (`HIST_K_POW2`). We expect the **compiler** to reduce 2-D masked tiles correctly for all `K_POW2`, not require padding to 8 int32 lanes.

### Workaround (application side)

- Pad inner `tl.arange` to at least **8** for histogram-style 2-D reductions while keeping `k < K` load masks, **or**
- Use 1-D flatten load + atomic (verified correct; older comment cited UB/D-cache concerns on long MoE runs—needs re-validation if chosen as permanent fix).

Neither should be mandatory for **`K_POW2 ∈ {2,4}`** if 2-D reductions are supported.

### Environment details

```
Triton: 3.2.0
triton_ascend: 3.2.2 (pip package name triton_ascend)
PyTorch: 2.9.0+cpu
torch_npu: 2.9.0.post4
NPU: Ascend950PR (reported device name: Ascend950PR_9579)
OS: Linux (openEuler 24.03 SP3, kernel 6.6.x)
Python: 3.11
```

Compile path: 
A5 — `bishengir-compile-a5 --target=Ascend950PR_9579` (default triton-ascend A5 pipeline)

Cross-SOC: 

Same histogram logic with `K=2` counts correctly on Ascend A3 in our downstream tests; incorrect totals on A5 as above.

Repro command:
```
  python repro_moe_histogram.py
  # compiles and launches; compare printed totals to ref=1024
```

**Comment by hongziqi** [comment]
Thanks for your feedback. We will arrange relevant engineers to investigate this issue later.

**Comment by hongziqi** [comment]
On Ascend950, `hist_2d_atomic` returns **256** instead of expected **1024** (`T*K`) when `INNER_POW2=2`. It works correctly with `INNER_POW2=8`.

Triton DSL and ttadapter IR are correct. The error occurs inside the outlined vector function when computing `valid = flat_m & (flat_e >=0) & (flat_e < E)`. 

Please consider forwarding this to the AscendNPU-IR community:
https://gitcode.com/Ascend/AscendNPU-IR


### Verified correct parts

ttadapter IR produces proper mask, reshape and bitwise AND.
NPU-IR entry: mask densified via `hivm.hir.copy` + `memref.collapse_shape` into `memref<1024xi1, ub>`, then passed to outlined VF.

**Runtime GM dump**:

| signal | nnz |
|--------|-----|
| densified `flat_m` alone | **1024** |
| `(flat_e >= 0) & (flat_e < E)` alone | **1024** |
| `valid = flat_m & cmp` | **256** |

### Suspected lowering bug in `hist_2d_atomic_fused_1_outlined_vf_0`
In the late intrinsic IR (`module.hivm.opt.mlir`), the VF loads `i32` ids with `vldsx1`, but loads the densified `memref<1024xi1>` mask via a different path (`udiv` by 8, `vldus.post.s32`, `movvp`) before `pand` with the compares:

```mlir
func.func @hist_2d_atomic_fused_1_outlined_vf_0(
    %arg0: memref<1024xi1, ub>,    // flat_m
    %arg1: memref<1024xi32, ub>,   // flat_e
    ...)

// mask load
%23 = llvm.udiv %22, %c8 : i64
%26 = hivm.vldus.post.s32(...)
%28 = hivm.movvp(...) -> vector<256xi1>

// ids load + compares look normal
%32 = hivm.vldsx1.v64s32(...)
%34 = hivm.vcmp.lt ...   // ids < E
%36 = hivm.vcmp.ge ...   // ids >= 0
%38 = hivm.pand.z(%28, %36, ...)
%40 = hivm.pand.z(%38, %34, ...)
```

We **suspect** that this VF lowering mis-handles the densified `i1` operand (so `pand` sees a wrong predicate).

**Comment by zheliuyu** [comment]
@hongziqi Thank you for pinpointing this issue. I've reported it to the AscendNPU-IR community: https://gitcode.com/Ascend/AscendNPU-IR/issues/603. Waiting for their response.

Let's keep this issue open for now, since the root cause hasn't been pinpointed yet.
